### DISCUSSION TITLE
Raw BNF grammars instead of Guidance grammar DSL

### DISCUSSION THREAD

**Discussion by a3lem** [guidance-ai/guidance#793] [Q&A]
Context:

I'm slowly wrapping my head around guidance's way of working and evaluating whether it's worth the added layer of abstraction.

One thing that strikes me as odd in the examples – although this might just say something about my current level of understanding – is that CFG grammars are consistently defined using Guidance operators, whereas BNF notation already exists. By contrast, regex patterns are defined using the familiar regex DSL. 

I suppose there are advantages to using a custom grammar DSL in terms of getting values from the generated output. I'm concerned, though, that I may at some point encounter a use case that is easier/only possible to define in BNF notation, which obviously factors into how I evaluate Guidance.

Questions:

1. In case I missed something, is it in fact already possible to provide BNF grammars directly?
2. If not, what was the reasoning for not including this?
3. Does the Guidance grammar DSL provably cover all possible expressions that BNF notation could define?
4. If not, would it be an idea to add direct support for BNF notation?

There's a good chance my questions simply reveal I haven't fully understood the power of Guidance yet. Anyway, thanks in advance for any extra info!

**Comment by riedgar-ms** [comment]
We are looking into EBNF support within `guidance` right now. Tag @hudson-ai who is experimenting with this.

As to 'why not the default?' the reason is that `guidance` tries to make the on-ramp for novice users as gentle as possible, and EBNF grammars are not well known outside of the field of computer science.

**Reply by hudson-ai** [reply]
@lemontheme if you're curious or want to play around with it a bit, here is my draft (really a draft -- the API and implementation are subject to change) of EBNF support for guidance, using `lark` as the EBNF parser:

https://github.com/hudson-ai/guidance/blob/ebnf/guidance/library/_ebnf.py

To add to Richard's comment on 'why not the default', I feel that guidance is doing a bit more than just constraining LLMs with CFGs. In my opinion, it's a pythonic interface that makes working with LLMs 'feel like programming' again, if that makes any sense. For loops, branches/if statements, function calling, etc. are all part of this interface.

That being said, I agree that some subsets of this functionality are most easily described with (E)BNF, hence this draft implementation. There are a few bumps in the road I want to smooth out before opening a PR, but I hope to get there soon :)

**Reply by hudson-ai** [reply]
See the tests here for an example of usage:

https://github.com/hudson-ai/guidance/blob/ebnf/tests/library/test_ebnf.py

**Reply by a3lem** [reply]
Interesting! Glad to see others have been thinking about this too. 

Your take on what Guidance aims to be got me thinking. Since posting, I've come to realize it's likely that Guidance, with its operators and imperative approach, is strictly more expressive than CFGs (although I don't know enough formal language theory to actually prove this). In other words, everything that can be generated from a CFG can be expressed in Guidance.

Still, when all one cares about is capturing a string generated from a complex grammar, then BNF is a whole lot more compact than defining tens of Guidance functions. Presumably, these are the subsets of functionality you refer to?

I'm still figuring out how your implementation works. It's forcing me to dig into the internal API. My high-level understanding is that convert a Lark grammar (+1 for Lark by the way) into internal Guidance concepts, which are then evaluated as usual. (There's a few steps in between that I'm still wrapping my head around. The Guidance codebase takes some time to understand, it seems.) Anyway, is that more or less what's happening?

**Reply by hudson-ai** [reply]
I also suspect that guidance is strictly more expressive than CFGs (although I also haven't seen nor produced a proof of this yet).

>  Still, when all one cares about is capturing a string generated from a complex grammar, then BNF is a whole lot more compact than defining tens of Guidance functions. Presumably, these are the subsets of functionality you refer to?

Yes, exactly!

I want to make one more comment on this part of the discussion that is more-or-less orthogonal to the formal "expressiveness" of the grammar, and it is related to my use of the phrase "pythonic interface" above. Even if guidance were "only" as expressive as BNF, I believe that its design comes with several valuable [affordances](https://en.wikipedia.org/wiki/Affordance) (both implicit and explicit) that will impact what kinds of systems people will be able to dream up and build.

Some examples to illustrate what I mean (heavily quoting a conversation I had with @notarealdeveloper earlier -- tagging you in case you are interested in weighing in):
1. You may choose to use one programming language over another not because it is more powerful (they are both probably Turing complete...) but because it makes it (cognitively) easier to implement the kind of functionality you are designing.
2. Choice of mathematical notation matters even when competing notations describe "the same thing". A good notation (where "good" is context dependent!) "tells" you what kinds of operations you can or should apply. See Terence Tao's excellent [mathoverflow post about this](https://mathoverflow.net/questions/366070/what-are-the-benefits-of-writing-vector-inner-products-as-langle-u-v-rangle/366118#366118).
3. Plenty of mathematical theorems are "equivalent" to each other (e.g. [Bolzano-Weierstrass and the statement that "every countable commutative ring has a maximal ideal"](https://en.wikipedia.org/wiki/Reverse_mathematics#Arithmetical_comprehension_ACA0)), but they _feel_ different. This "feeling different" goes beyond "vibes" and has real-world consequences. If "mathematical equivalence" and "cognitive equivalence" were the same, then nearly all mathematical theorems would be provable with zero work.

It's my hope that the language/interface that guidance provides is more suitable and approachable than BNF for programmers to design the kinds of grammars that will be truly useful to them when applying LLMs to real-world problems. The only way to get there is via community feedback though, so thank you so much for engaging here!

Anyway, back to your questions/comments about the implementation. Guidance essentially has two primitives: "select" and "join", where select is equivalent to the "or" (`|`) operator in EBNF and join is just concatenation/juxtaposition. To make a guidance grammar from an EBNF specification, I'm just taking advantage of this equivalence. So in other words, yes -- you hit the nail on the head about how this works.

I'll admit that my implementation is pretty opaque at the moment... I believe this is because of some subtleties around how guidance handles recursive calls of guidance-decorated functions as well as the difficulties that present themselves when programmatically generating code. I'm hoping to simplify the implementation a bit, but hopefully it should be usable without you having to understand all of the internals anyway :smile:

Btw thanks for the +1 for Lark -- that's a valuable signal! Do you know if there are other competing or complimentary libraries that may be worth considering here?

**Reply by a3lem** [reply]
Thanks for introducing me to several new concepts! That's a fascinating way of looking at it. Things got philosophical there in the middle, but I'm here for it!

While some of Guidance's more advanced functionality has taken me a while to wrap my head around, I think it's a fact that, out of many LLM prompting libraries, Guidance offers the most direct route to just having an actionable 'conversation' with an LLM – 'actionable' in the sense that a minimum of new abstractions is needed to get output from one generation step, make a decision, and pass it to the next generation step.

I'm still on the fence about where the future is headed. I'd love to see a DSL that can be compiled to something that can be run hyper-efficiently in an LLM runtime. But then, tack on enough features to a DSL and soon enough it becomes Turing complete, at which point you might as well have just used an existing programming language – which can similarly be compiled for efficient execution, except without the extra step in the middle. =p 

Your implementation isn't necessarily opaque, to be clear. It's the whole codebase. A question I find particularly hard to answer is _Where does the generation call happen?_ – because something I find essential to check when reviewing any of the hundreds of emerging LLM prompting libraries is that the library is not modifying my prompts/messages. I'm still learning how everything fits together. (Something I'm looking to implement/add is a way to use a Guidance program both for prompting/inference (i.e. answer is unknown) and for generating training examples (answer is known, but the format should resemble what will be used at inference time).)

As for Lark, there are a few libraries like it, obviously. Once upon a time, I reviewed all of them, settling on Lark in the end. I can't remember why. I think I liked the notation flavor and the ability to import grammars; found the docs clear; the speed, adequate; oh, and that it didn't bring any extra dependencies.

**Comment by hudson-ai** [comment]
> Things got philosophical there in the middle, but I'm here for it!

I tend to get a bit rambly at times, so it's nice to hear that it's appreciated :laughing: 

> Guidance offers the most direct route to just having an actionable 'conversation' with an LLM – 'actionable' in the sense that a minimum of new abstractions is needed to get output from one generation step, make a decision, and pass it to the next generation step.

:heart: 

> I'd love to see a DSL that can be compiled to something that can be run hyper-efficiently in an LLM runtime. But then, tack on enough features to a DSL and soon enough it becomes Turing complete, at which point you might as well have just used an existing programming language – which can similarly be compiled for efficient execution, except without the extra step in the middle. =p

That is essentially the goal here! *You* write python, *we* turn that into something that can be used to :crossed_fingers: efficiently guide an LLM.

> Where does the generation call happen?

Great question. Hopefully I can help address it a bit...

When you call a guidance function, _nothing really happens_. That's imprecise, but what I mean is that all that really goes on under the hood is that the python code inside the function is used to generate a grammar (sort of like a compilation step). Functions that require dynamic python control-flow are a little more complicated than that, but that's the general idea (this is the distinction between "stateless" and "stateful" grammars that you can find in the readme). 

Adding strings and/or the grammars returned by guidance functions (and strings) together simply concatenates the strings/grammars. Still no LLM calls. 

Generation only happens when one of these grammars is added to the right of a `Model` object. You can see this in the `__add__` method on the `Model` class. Most of the actual work is being done in the `_run_stateless` method, which accepts one of these "compiled" grammars and runs the LLM forward on it. The implementation is a bit opaque, but hopefully just knowing _where it happens_ provides some insight.

Your prompts should never really be "modified" by this library (beyond possibly some chat/role blocks, which I believe are more or less "opt-in"). Adding a string to an LLM is quite literally appending that string to the "prompt" used for further generation. Adding a grammar just runs the LLM forward on that grammar, furthermore appending the output of that generation to the downstream "prompt". I'm putting quotes on "prompt" because I think a nice insight of this library is that there really isn't a distinction between prompts and deterministic generation constraints.

**Comment by ibehnam** [comment]
Just as as follow-up, I wanted to see if there's been any update on this? It's so cool to be able to use (E)BNF grammars directly in guidance. @hudson-ai

**Reply by hudson-ai** [reply]
Hi @ibehnam sorry for the long delay! I have an open PR on this at the moment. DEFINITELY not going to be a stable/public part of the guidance API for a little while, but if you want to play around with it, I would love to get some feedback. See the PR here: https://github.com/guidance-ai/guidance/pull/1017

See `guidance/library/_rfc3986.py` for an example usage

**Reply by nchammas** [reply]
> I have an open PR on this at the moment. DEFINITELY not going to be a stable/public part of the guidance API for a little while, but if you want to play around with it, I would love to get some feedback. See the PR here: #1017

It looks like this work got folded into the LLGuidance project, but I'm not clear on whether or not it's still not ready for public use.

What, for example, is the current equivalent to [this test of EBNF grammar support][1] that you wrote?

[1]: https://github.com/hudson-ai/guidance/blob/67fcdd1b197f7af9bb92a8dcd9e154c18249f07c/tests/library/test_ebnf.py#L9

**Reply by hudson-ai** [reply]
@nchammas hey, thanks for pushing on this! We actually just exposed support for lark-style EBNF today. I'd still call it experimental for sure, but we're pretty firm on the lark-like syntax and the related semantics [discussed in these docs here](https://github.com/guidance-ai/llguidance/blob/main/docs/syntax.md). You are indeed correct that it ended up being folded into `llguidance`. 

See my comment here for some indication on usage in guidance:
https://github.com/guidance-ai/guidance/issues/1135#issuecomment-2711199256

Related example to that test EBNF:

```lark
lark_grm = """
start: expression

expression: INTEGER
     | expression OPERATOR expression
     | "(" expression ")"

INTEGER: "-"? DIGIT+
DIGIT: /[0-9]/
OPERATOR: "+"
     | "*"
     | "**"
     | "/"
     | "-"

%ignore /\s/
"""
```

Let me know if you run into any difficulties 😄

**Reply by nchammas** [reply]
@hudson-ai - Can you share the exact version or git hash you are running for guidance, llguidance, transformers, and/or llama-cpp-python, plus the model you are using? Asking because there are so many broken combinations right now, and I figure it's easier for me to just ask what you are using than to keep trying random combinations.

For the record, this is what I have installed:

```
guidance                0.2.0 514a5eb (i.e. master as of 2025-03-13)
llama-cpp-python        0.3.8
llguidance              0.6.31
torch                   2.6.0
transformers            4.49.0
```

Trying a basic Lark grammar as follows:

```python
grammar_def = """
expr    : expr "+" term     -> add
        | expr "-" term     -> sub
        | term
term    : term "*" factor   -> mul
        | term "/" factor   -> div
        | factor
factor  : integer
        | "(" expr ")"
integer : DIGIT+
        | "-" integer       -> neg
%import common.DIGIT
"""
grammar = lark(grammar_def)
model = models.Transformers("microsoft/Phi-4-mini-instruct")
print(model + grammar)
```

Results in an `EOFError`, similar to what I [reported here][1].

[1]: https://github.com/guidance-ai/guidance/issues/1151#issuecomment-2707261384

**Reply by hudson-ai** [reply]
@nchammas what OS are you using? The example you sent (with one modification -- we need an explicit `start` rule, e.g. `start: expr`) works for me on macOS Sequoia 15.3.1:

```
guidance             0.2.0   (514a5eb)
llama-cpp-python     0.3.8
llguidance           0.7.0   (works on 0.6.31 for me as well)
torch                2.6.0
transformers         4.49.0
```

**Reply by hudson-ai** [reply]
Ah, hold up! I am able to reproduce when running the example code as a script (it works fine interactively in `ipython`). @nopdive it looks like the metrics monitoring code is spawning a subprocess that breaks when run non-interactively. Would you be able to take a look?

**Reply by nchammas** [reply]
I am using macOS 15.3.2 and, yes, I am running things on Python 3.11.11 via `python example.py` and not in a notebook.
